packages/skygear-core/lib/cloud/transport/common.js

```diff
@@ -104,27 +104,130 @@ class SkygearRequest {
  * interface for setting response headers and body.
  */
 export class SkygearResponse {
-  constructor(options) {
+  /**
+   * Creates an instance of SkygearResponse.
+   *
+   * @param {Object} [options={}] - options to initialize the response
+   * @param {number} [options.statusCode=200] - HTTP status code of the response
+   * @param {string} [options.body=''] - HTTP response body
+   * @param {Object} [options.headers={}] - HTTP response headers
+   */
+  constructor(options = {}) {
+    /**
+     * The HTTP status code of the response.
+     *
+     * @type {number}
+     */
     this.statusCode = options.statusCode || 200;
+
+    /**
+     * The HTTP headers of the response.
+     *
+     * @type {Object}
+     */
     this.headers = options.headers || {};
+
+    /**
+     * The HTTP body of the response.
+     *
+     * @type {string}
+     */
     this.body = options.body || '';
   }
 
+  /**
+   * Set a HTTP header to the response.
+   * 
+   * @param {string} name - HTTP header name
+   * @param {string} value - HTTP header value
+   */
   setHeader(name, value) {
     this.headers[name] = value;
   }
 
+  /**
+   * Get a HTTP header from the response.
+   * 
+   * @param {string} name - HTTP header name
+   * @return {string} HTTP header value
+   */
   getHeader(name) {
     return this.headers[name];
   }
 
+  /**
+   * Remove a HTTP header from the response.
+   * 
+   * @param {string} name - HTTP header name
+   */
   removeHeader(name) {
     delete this.headers[name];
   }
 
+  /**
+   * Write a chunk of data into the response. The chunk will be appended
+   * to any existing data in the response body.
+   * 
+   * @param {string} chunk - data to append to the response body
+   */
   write(chunk) {
     this.body += chunk;
   }
+
+  /**
+   * Convert the response to a result JSON that is suitable for plugin
+   * transport.
+   * 
+   * @return {Object} result JSON for plugin transport
+   */
+  toResultJSON() {
+    const header = {};
+    const status = this.statusCode || 200;
+    const body = b64EncodeUnicode(this.body);
+
+    Object.keys(this.headers).forEach((perKey) => {
+      header[perKey] = this.headers[perKey];
+    });
+
+    return {
+      header,
+      status,
+      body
+    };
+  }
+
+  /**
+   * Wrap response body into a SkygearResponse.
+   *
+   * If the specified value is a SkygearResponse, the same object will
+   * be returned.
+   * 
+   * @param result - SkygearResponse or response body
+   * @return {!SkygearResponse} a SkygearResponse
+   */
+  static wrap(result) {
+    if (result instanceof SkygearResponse) {
+      return result;
+    } else if (result.body !== undefined) {
+      return new SkygearResponse(result);
```

Comment:
> Wouldn't this make the behaviour strange?

Yes.

I need a way to recognize that the returned object from a handler is of type SkygearResponse. As far as I know, in JS, the way a JS is imported will affect the `result instanceof SkygearReponse` check, meaning that I am not able to accurately determine if the object in question is a SkygearResponse. A fallback to that is to check if the object has a `body` field.

(The case that returning a `SkygearResponse` from a handler but the transport not recognising that the object is of type `SkygearResponse` usually happens on the cloud. This is because the cloud runs `npm install skygear` and `npm install`, so it is possible to end up with two copies of `SkygearResponse` class.)

I am open to suggestion on how best to accomplish this.